File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @param {number} k
     * @return {ListNode}
     */
    reverseKGroup(head, k) {
        let current = head;
        let groupCount = 0;
        while (current && groupCount < k) {
            current = current.next;
            ++groupCount;
        }


        if (groupCount === k) {
            current = this.reverseKGroup(current, k);
            while (groupCount > 0) {
                let tmp = head.next;
                head.next = current;
                current = head;
                head = tmp;
                groupCount--;

            }

            head = current;
        }

        return head;

    }
}

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode i = head, j = head;
        while (i != null && i.next != null) {
            i = i.next.next;
            j = j.next;
            if (i == j) return true;
        }
        return false;
    }
}
